src/ergw_aaa_diameter.erl

```diff
@@ -333,75 +313,102 @@ framed_protocol('GPRS-PDP-Context')        -> ?'DIAMETER_SGI_FRAMED-PROTOCOL_GPR
-       'Accounting-Input-Octets' = attr_get('InOctets', Session),
-       'Accounting-Input-Packets' = attr_get('InPackets', Session),
-       'Accounting-Output-Octets' = attr_get('OutOctets', Session),
-       'Accounting-Output-Packets' = attr_get('OutPackets', Session),
-       'Acct-Authentic' = [acct_authentic(attr_get('Acct-Authentic', Session))],
-       'Called-Station-Id' = attr_get('Called-Station-Id', Session),
-       'Calling-Station-Id' = attr_get('Calling-Station-Id', Session),
-       'Framed-Interface-Id' = attr_get('Framed-Interface-Id', Session),
-       'Framed-IP-Address' = attr_get_addr('Framed-IP-Address', Session),
-       'Framed-Protocol' = [framed_protocol(attr_get('Framed-Protocol', Session))],
-       'Tunneling' = tunneling(Session, State),
-       '3GPP-IMSI' = attr_get('3GPP-IMSI', Session),
-       '3GPP-Charging-Id' = attr_get('3GPP-Charging-ID', Session),
-       '3GPP-PDP-Type' = [pdp_type(attr_get('3GPP-PDP-Type', Session))],
-       '3GPP-CG-Address' = attr_get_addr('3GPP-Charging-Gateway-Address', Session),
-       '3GPP-GPRS-Negotiated-QoS-Profile' = attr_get('3GPP-GPRS-Negotiated-QoS-Profile', Session),
-       '3GPP-SGSN-Address' = attr_get_addr('3GPP-SGSN-Address', Session),
-       '3GPP-GGSN-Address' = attr_get_addr('3GPP-GGSN-Address', Session),
-       '3GPP-IMSI-MCC-MNC' = attr_get('3GPP-IMSI-MCC-MNC', Session),
-       '3GPP-GGSN-MCC-MNC' = attr_get('3GPP-GGSN-MCC-MNC', Session),
-       '3GPP-NSAPI' = octet_string(attr_get('3GPP-NSAPI', Session)),
-       '3GPP-Selection-Mode' = int_to_binary(attr_get('3GPP-Selection-Mode', Session)),
-       '3GPP-Charging-Characteristics' = format_cc(attr_get('3GPP-Charging-Characteristics', Session)),
-       '3GPP-CG-IPv6-Address' = attr_get_addr('3GPP-Charging-Gateway-IPv6-Address', Session),
-       '3GPP-SGSN-IPv6-Address' = attr_get_addr('3GPP-SGSN-IPv6-Address', Session),
-       '3GPP-GGSN-IPv6-Address' = attr_get_addr('3GPP-GGSN-IPv6-Address', Session),
-       '3GPP-SGSN-MCC-MNC' = attr_get('3GPP-SGSN-MCC-MNC', Session),
-       '3GPP-IMEISV' = attr_get('3GPP-IMEISV', Session),
-       '3GPP-RAT-Type' = octet_string(attr_get('3GPP-RAT-Type', Session)),
-       '3GPP-User-Location-Info' = attr_get('3GPP-User-Location-Info', Session),
-       '3GPP-MS-TimeZone' = octet_string(attr_get('3GPP-MS-TimeZone', Session)),
-       '3GPP-CAMEL-Charging-Info' = attr_get('3GPP-Camel-Charging', Session),
-       '3GPP-Packet-Filter' = attr_get('3GPP-Packet-Filter', Session),
-       '3GPP-Negotiated-DSCP' = attr_get('3GPP-Negotiated-DSCP', Session),
-       'AVP' = []}.
-
-% TODO: build #'Tunneling'{}
-tunneling(_Session, _State) -> [].
+    Acc = [{'Session-Id', State#state.sid},
+           {'Accounting-Record-Type', Type},
+           {'Accounting-Record-Number', State#state.accounting_record_number},
+           {'NAS-Identifier', [State#state.nas_id]}],
+    ['ACR' | lists:foldl(fun({Key, Value}, Acc0) ->
```

Comment:
the pcc_rules branch is doing this to other way around, it first constructs a DIAMTER request as a map by walking the session map and applying the formatting function. If there is no formatting/conversion for a element from the session it is simply skipped. In a second step it then converts the request from map to list.

You're version converts the session map into a list (copying all elements) and the applies the formatting function to all members of the list. This only works if the DIAMTER request has a flat structure (or a structure that is a direct translation of the session object), if a more complicated mapping is required, this will result in quite a mess.

So, to keep the pcc rules branch in sync with this change, please use the method from the pcc rules branch.